Start chat when user adds a pending contact from the "new contact" view

cammellos opened the pull request:
fixes #2960 

### Summary:

At the moment removing a contact hides it, setting the pending? flag to true.

If you remove and add the contact again from the "new contact" screen, the user is not sent to the 1-1 chat.

I have changed the code so that :start-chat is dispatched regardless whether the user is new or was already in the contacts.

### Steps to test:
- Open Status
- Remove a contact
- Copy their public key
- Go back to Home
- Go to Add a contact
- Insert their public key & click on add

status: ready

cammellos committed:
Start chat when user adds a pending contact from the "new contact" view

When a contact is removed it is set to pending?.

If the same contact is added again we now dispatch a :start-chat event
which initiate a 1-1 chat with them.

goranjovic commented on src/status_im/ui/screens/contacts/events.cljs:
```diff
@@ -437,8 +437,9 @@
-                                                    :whisper-identity id}])})))
+      {:dispatch-n (if (get-in db [:contacts/contacts id])
+                     [[:add-pending-contact id]
+                      [:start-chat id {:navigation-replace? true}]]
```
Maybe better to avoid dispatch-n and just have one event that both adds pending contact and start the chat in the true case.

goranjovic commented on src/status_im/ui/screens/contacts/events.cljs:
```diff
@@ -437,8 +437,9 @@
-                                                    :whisper-identity id}])})))
+      {:dispatch-n (if (get-in db [:contacts/contacts id])
+                     [[:add-pending-contact id]
+                      [:start-chat id {:navigation-replace? true}]]
```
Maybe better to avoid dispatch-n and just have one event that both adds pending contact and start the chat in the true case.

cammellos replied:
cool,

just to be sure,

you mean something like `:add-pending-contact-and-open-chat`?
Basically in similar fashion as `:add-new-contact-and-open-chat`.

Thanks

goranjovic replied:
Yes, exactly. Cheers!

cammellos commented:
cool,

just to be sure, 

you mean something like `:add-pending-contact-and-open-chat `?
Basically in similar fashion as `:add-new-contact-and-open-chat`.

Thanks

cammellos committed:
Add :add-pending-contact-and-open-chat handler

cammellos commented:
I have created a new handler `:add-pending-contact-and-open-chat`

I could not replace `:add-pending-contact` as it is used in other places where we don't want to open the chat screen: 

https://github.com/status-im/status-react/blob/develop/src/status_im/chat/screen.cljs#L43 

and 

https://github.com/status-im/status-react/blob/2256628c97897c05064f8eb57687f69e2ca41c97/src/status_im/ui/screens/profile/views.cljs#L71 

I took a similar approach to:

https://github.com/status-im/status-react/blob/2256628c97897c05064f8eb57687f69e2ca41c97/src/status_im/ui/screens/accounts/events.cljs#L131

as this is a pattern used before.

We could also extend `:add-pending-contact` to conditionally dispatch the `:start-chat` event, somehow similarly to:

https://github.com/status-im/status-react/blob/7685b430494025dc58ac72b0510f4eb04f207a1a/src/status_im/ui/screens/events.cljs#L274

but the former approach seems simpler.

cheers

janherich commented on src/status_im/ui/screens/contacts/events.cljs:
```diff
@@ -346,6 +346,12 @@
+  :add-pending-contact-and-open-chat
+  (fn [_ [_ whisper-id]]
+    {:dispatch-n [[:add-pending-contact whisper-id]
+                  [:start-chat whisper-id {:navigation-replace? true}]]}))
```
Please, instead of delegating dispatches in this (now atomic handler), compose the functions instead.
For starting chat, the function producing necessary re-frame effects is here - https://github.com/status-im/status-react/blob/develop/src/status_im/chat/events.cljs#L298-L306 (it's anonymous function currently, now it's right time to make it named and re-use it :) )
For adding pending contact, it's here - https://github.com/status-im/status-react/blob/develop/src/status_im/ui/screens/contacts/events.cljs#L337-L347 (anonymous function again, and worse then starting chat, because its unnecessary dispatching events instead of composing functions).
See this for rationale - https://github.com/status-im/ideas/issues/31#issuecomment-346345981 (point 3.)

flexsurfer replied:
and also please check you should go back to home from chat screen not to add new contact

cammellos replied:
@janherich 
just to make sure we are on the same page, you mean something like:

https://github.com/cammellos/status-react/commit/bb8c6f06dd44bf302263a4f542d1e249d608234c#diff-0da314cfa3a44d32c27ef25c73fca7cbR320

eventually replacing the last `:dispatch` as well (`::send-contact-request`)

( I started with `add-new-contact` for simplicity)

janherich replied:
@cammellos Exactly, very good work !

cammellos commented:
thanks for the feedback, I will take a look at it

cheers

On Jan 14, 2018 9:26 PM, "Jan Herich" <notifications@github.com> wrote:

> *@janherich* commented on this pull request.
> ------------------------------
>
> In src/status_im/ui/screens/contacts/events.cljs
> <https://github.com/status-im/status-react/pull/3031#discussion_r161410595>
> :
>
> > @@ -346,6 +346,12 @@
>                      [:watch-contact contact']
>                      [:discoveries-send-portions chat-or-whisper-id]]})))
>
> +(register-handler-fx
> +  :add-pending-contact-and-open-chat
> +  (fn [_ [_ whisper-id]]
> +    {:dispatch-n [[:add-pending-contact whisper-id]
> +                  [:start-chat whisper-id {:navigation-replace? true}]]}))
>
> Please, instead of delegating dispatched in this (now atomic handler),
> compose the function instead.
> For starting chat, the function producing necessary re-frame effects is
> here - https://github.com/status-im/status-react/blob/develop/src/
> status_im/chat/events.cljs#L298-L306 (it's anonymous function currently,
> , now it's right time to make it named and re-use it :) )
> For adding pending contact, it's here - https://github.com/status-im/
> status-react/blob/develop/src/status_im/ui/screens/contacts/
> events.cljs#L337-L347 (anonymous function again, and worse then starting
> chat, because its unnecessary dispatching events instead of composing
> functions).
> See this for rationale - status-im/ideas#31 (comment)
> <https://github.com/status-im/ideas/issues/31#issuecomment-346345981>
> (point 3.)
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/status-im/status-react/pull/3031#pullrequestreview-88694955>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AA-EsImQepure9mGRjND_L5kxin1c8Ubks5tKnDqgaJpZM4RdYur>
> .
>

flexsurfer commented on test/cljs/status_im/test/contacts/events.cljs:
```diff
@@ -152,7 +154,7 @@
    load-default-contacts (add-contact-groups, add-contacts, add-contacts-to-group ;TODO add-chat, load-commands!)
    add-contact-handler (add-new-contact-and-open-chat, status-im.contacts.events/add-new-contact,
-                        status-im.contacts.events/send-contact-request ;TODO start-chat)
+                        status-im.contacts.events/send-contact-request, status-im.chat.events.start-chat)
```
status-im.chat.events **/** start-chat ?

cammellos replied:
Probably I misunderstood the comment, 
I thought that it meant that the tests were implicitly testing this event: `status-im.chat.events.start-chat`

cammellos committed:
Refactor add-contacts & add-pending-contact

We wanted to move towards having multiple functions
manipulating the data rather than relying on a chain of
dispatched events.

I have refactored ui.screens.contacts.events so that now the
functions mainly manipulate fx, passed as a first parameter.

Their responsability is to make sure their fx are merged in, so
that they can be composed more easily.

Signed-off-by: Eric Dvorsak <eric@dvorsak.fr>

cammellos commented:
I have made some changes as discussed,

basically I have changed some of the functions in the contacts/events namespace to act on the fx map.

They have the responsibility of merging the new effects and returning the fx map, in this way it's easy to compose/thread functions.

Let me know what you think.
cheers

cammellos commented:
Also, not sure why it's not being build on Jenkins, any idea?

goranjovic commented:
@cammellos It's a problem with Jenkins. We'll sort it out.

oskarth commented:
@janherich @goranjovic @flexsurfer If this is has passed review, please move to test.

Serhy commented:
Hey @cammellos, could you please rebase this branch of develop?

cammellos commented:
@Serhy done
cheers

Serhy commented:
Tested on build c463e487+ for Android and 0.0.2163 on iOS with various cases. 
All good! In any case of adding contact from 'Add new contact' screen, - if this contact not in the list - I wind up on 1to1 chat after public key (via scanned QR code or paste the public key value) submitted. 
Ready for merge!
